File: app/src/debug/java/org/tasks/BuildSetup.java
package org.tasks;

import android.app.Application;
import android.content.Context;
import android.os.StrictMode;
import com.facebook.flipper.android.AndroidFlipperClient;
import com.facebook.flipper.android.utils.FlipperUtils;
import com.facebook.flipper.core.FlipperClient;
import com.facebook.flipper.plugins.databases.DatabasesFlipperPlugin;
import com.facebook.flipper.plugins.inspector.DescriptorMapping;
import com.facebook.flipper.plugins.inspector.InspectorFlipperPlugin;
import com.facebook.flipper.plugins.network.NetworkFlipperPlugin;
import com.facebook.flipper.plugins.sharedpreferences.SharedPreferencesFlipperPlugin;
import com.facebook.soloader.SoLoader;
import javax.inject.Inject;
import leakcanary.AppWatcher;
import org.tasks.injection.ForApplication;
import org.tasks.preferences.Preferences;
import timber.log.Timber;

class BuildSetup {

  private final Context context;
  private final Preferences preferences;

  @Inject
  BuildSetup(@ForApplication Context context, Preferences preferences) {
    this.context = context;
    this.preferences = preferences;
  }

  public boolean setup() {
    Timber.plant(new Timber.DebugTree());
    Application application = (Application) context.getApplicationContext();
    SoLoader.init(application, false);

    if (!preferences.getBoolean(R.string.p_leakcanary, false)) {
      AppWatcher.setConfig(AppWatcher.getConfig().newBuilder().enabled(false).build());
    }
    if (preferences.getBoolean(R.string.p_flipper, false) && FlipperUtils.shouldEnableFlipper(context)) {
      FlipperClient client = AndroidFlipperClient.getInstance(application);
      client.addPlugin(new InspectorFlipperPlugin(application, DescriptorMapping.withDefaults()));
      client.addPlugin(new DatabasesFlipperPlugin(application));
      client.addPlugin(new NetworkFlipperPlugin());
      client.addPlugin(new SharedPreferencesFlipperPlugin(application));
      client.start();
    }
    if (preferences.getBoolean(R.string.p_strict_mode_thread, false)) {
      StrictMode.setThreadPolicy(
          new StrictMode.ThreadPolicy.Builder()
              .detectDiskReads()
              .detectDiskWrites()
              .detectNetwork()
              .penaltyLog()
              .build());
    }
    if (preferences.getBoolean(R.string.p_strict_mode_vm, false)) {
      StrictMode.setVmPolicy(
          new StrictMode.VmPolicy.Builder()
              .detectLeakedSqlLiteObjects()
              .detectLeakedClosableObjects()
              .penaltyLog()
              .build());
    }
    return true;
  }
}

File: app/src/debug/java/org/tasks/BuildSetup.kt
package org.tasks

import android.app.Application
import android.os.StrictMode
import android.os.StrictMode.VmPolicy
import com.todoroo.andlib.utility.AndroidUtilities.atLeastOreo
import com.todoroo.andlib.utility.AndroidUtilities.atLeastQ
import leakcanary.AppWatcher
import org.tasks.logging.FileLogger
import org.tasks.preferences.Preferences
import timber.log.Timber
import javax.inject.Inject

class BuildSetup @Inject constructor(
    private val context: Application,
    private val preferences: Preferences,
    private val fileLogger: FileLogger,
) {
    fun setup() {
        Timber.plant(Timber.DebugTree())
        Timber.plant(fileLogger)
        if (preferences.getBoolean(R.string.p_leakcanary, false)) {
            AppWatcher.manualInstall(context)
        }
        if (preferences.getBoolean(R.string.p_strict_mode_thread, false)) {
            val builder = StrictMode.ThreadPolicy.Builder().detectAll().penaltyLog()
            if (preferences.getBoolean(R.string.p_crash_main_queries, false)) {
                builder.penaltyDeath()
            }
            StrictMode.setThreadPolicy(builder.build())
        }
        if (preferences.getBoolean(R.string.p_strict_mode_vm, false)) {
            val builder = VmPolicy.Builder()
                    .detectActivityLeaks()
                    .detectLeakedSqlLiteObjects()
                    .detectLeakedRegistrationObjects()
                    .detectLeakedClosableObjects()
                    .detectFileUriExposure()
                    .penaltyLog()
            if (atLeastOreo()) {
                builder.detectContentUriWithoutPermission()
            }
            if (atLeastQ()) {
                builder
                        .detectCredentialProtectedWhileLocked()
                        .detectImplicitDirectBoot()
            }
            StrictMode.setVmPolicy(builder.build())
        }
    }
}
